### ISSUE TITLE
Explicit jaxlib dependency causing incompatible environments

### ISSUE THREAD

**Issue by mattwthompson** [conda-forge/pymbar-feedstock#30]
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

The explicit `jaxlib` dependency is causing incompatible packages in some environments. There are warts in how JAX is packaged to conda, some of which result in the possibility of old (< 0.3) `jaxlib` builds sneaking in. One example is [here](https://github.com/openforcefield/openff-evaluator/actions/runs/4408343406/jobs/7723224426). I can't reproduce it locally, but it reliably happens in CI and causes tests to crash, so it stands to reason some users may be affected.

Some context can be found starting here: https://github.com/conda-forge/jax-feedstock/issues/81, but in summary: starting around March 2022 (version 0.3), `jaxlib`'s feedstock was updated to only install alongside versions of `jax` that were the same or greater. This fixed the issue moving forward but old builds were not removed, hence how I'm seeing old (0.1.x) versions of `jaxlib` alongside less-old (0.3) versions of `jax`. Previously, there were no constraints on which versions of each could be installed alongside each other.

I'm not sure why it is listed, it's not [imported at runtime](https://github.com/search?q=repo%3Achoderalab%2Fpymbar%20jaxlib&type=code) and the JAX feedstock [reliably depends](https://github.com/conda-forge/jax-feedstock/blob/main/recipe/meta.yaml) on `jaxlib`.

I can work around this but at some cost and complexity; users pulling down version 3 don't need JAX at all.

### Installed packages

```shell
# packages in environment at /home/runner/micromamba-root/envs/evaluator-test-env:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
absl-py                   1.4.0              pyhd8ed1ab_0    conda-forge
alabaster                 0.7.13             pyhd8ed1ab_0    conda-forge
alsa-lib                  1.2.8                h166bdaf_0    conda-forge
amberlite                 22.0                     pypi_0    pypi
ambertools                22.0             py38h5256d7f_3    conda-forge
soupsieve                 2.3.2.post1        pyhd8ed1ab_0    conda-forge
sphinx                    6.1.3              pyhd8ed1ab_0    conda-forge
sphinxcontrib-applehelp   1.0.4              pyhd8ed1ab_0    conda-forge
sphinxcontrib-devhelp     1.0.2                      py_0    conda-forge
sphinxcontrib-htmlhelp    2.0.1              pyhd8ed1ab_0    conda-forge
sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
sphinxcontrib-qthelp      1.0.3                      py_0    conda-forge
sphinxcontrib-serializinghtml 1.1.5              pyhd8ed1ab_2    conda-forge
sqlalchemy                2.0.5.post1      py38h1de0b5d_0    conda-forge
stack_data                0.6.2              pyhd8ed1ab_0    conda-forge
taproom                   0.2                pyh9f0ad1d_0    conda-forge
tblib                     1.7.0              pyhd8ed1ab_0    conda-forge
terminado                 0.17.1             pyh41d4057_0    conda-forge
tinycss2                  1.2.1              pyhd8ed1ab_0    conda-forge
tk                        8.6.12               h27826a3_0    conda-forge
toml                      0.10.2             pyhd8ed1ab_0    conda-forge
tomli                     2.0.1              pyhd8ed1ab_0    conda-forge
toolz                     0.12.0             pyhd8ed1ab_0    conda-forge
tornado                   6.2              py38h0a891b7_1    conda-forge
tqdm                      4.65.0             pyhd8ed1ab_1    conda-forge
traitlets                 5.9.0              pyhd8ed1ab_0    conda-forge
typing-extensions         4.5.0                hd8ed1ab_0    conda-forge
typing_extensions         4.5.0              pyha770c72_0    conda-forge
uncertainties             3.1.7              pyhd8ed1ab_0    conda-forge
unicodedata2              15.0.0           py38h0a891b7_0    conda-forge
urllib3                   1.26.15            pyhd8ed1ab_0    conda-forge
wcwidth                   0.2.6              pyhd8ed1ab_0    conda-forge
webencodings              0.5.1                      py_1    conda-forge
websocket-client          1.5.1              pyhd8ed1ab_0    conda-forge
wheel                     0.38.4             pyhd8ed1ab_0    conda-forge
widgetsnbextension        3.6.1              pyha770c72_0    conda-forge
xcb-util                  0.4.0                h166bdaf_0    conda-forge
xcb-util-image            0.4.0                h166bdaf_0    conda-forge
xcb-util-keysyms          0.4.0                h166bdaf_0    conda-forge
xcb-util-renderutil       0.3.9                h166bdaf_0    conda-forge
xcb-util-wm               0.4.1                h166bdaf_0    conda-forge
xkeyboard-config          2.38                 h0b41bf4_0    conda-forge
xmltodict                 0.13.0             pyhd8ed1ab_0    conda-forge
xorg-kbproto              1.0.7             h7f98852_1002    conda-forge
xorg-libice               1.0.10               h7f98852_0    conda-forge
xorg-libsm                1.2.3             hd9c2040_1000    conda-forge
xorg-libx11               1.8.4                h0b41bf4_0    conda-forge
xorg-libxau               1.0.9                h7f98852_0    conda-forge
xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge
xorg-libxext              1.3.4                h0b41bf4_2    conda-forge
xorg-libxrender           0.9.10            h7f98852_1003    conda-forge
xorg-libxt                1.2.1                h7f98852_2    conda-forge
xorg-renderproto          0.11.1            h7f98852_1002    conda-forge
xorg-xextproto            7.3.0             h0b41bf4_1003    conda-forge
xorg-xproto               7.0.31            h7f98852_1007    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
yaml                      0.2.5                h7f98852_2    conda-forge
yank                      0.25.2             pyhd8ed1ab_2    conda-forge
zeromq                    4.3.4                h9c3ff4c_1    conda-forge
zict                      2.2.0              pyhd8ed1ab_0    conda-forge
zipp                      3.15.0             pyhd8ed1ab_0    conda-forge
zlib                      1.2.13               h166bdaf_4    conda-forge
zstd                      1.5.2                h3eb15da_6    conda-forge
```

### Environment info

```shell
active environment : /home/runner/micromamba-root/envs/evaluator-test-env
    active env location : /home/runner/micromamba-root/envs/evaluator-test-env
            shell level : 1
       user config file : /home/runner/.condarc
 populated config files : /home/runner/.condarc
          conda version : 23.1.0
    conda-build version : not installed
         python version : 3.10.9.final.0
       virtual packages : __archspec=1=x86_64
                          __glibc=2.35=0
                          __linux=5.15.0=0
                          __unix=0=0
       base environment : /usr/share/miniconda  (writable)
      conda av data dir : /usr/share/miniconda/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/jaimergp/label/unsupported-cudatoolkit-shim/linux-64
                          https://conda.anaconda.org/jaimergp/label/unsupported-cudatoolkit-shim/noarch
                          https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /usr/share/miniconda/pkgs
                          /home/runner/.conda/pkgs
       envs directories : /usr/share/miniconda/envs
                          /home/runner/.conda/envs
               platform : linux-64
             user-agent : conda/23.1.0 requests/2.28.1 CPython/3.10.9 Linux/5.15.0-1034-azure ubuntu/22.04.2 glibc/2.35
                UID:GID : 1001:123
             netrc file : None
           offline mode : False
```

**Comment by Lnaden** [comment]
Well thats a frustrating bug, Thank you for finding this. I'm fine with #31 since the `jax` package explicitly needs `jaxlib` in its own recipe. JAX itself cannot run without jaxlib and will throw an import error if you try to just bring in jax without its lib installed.

> I'm not sure why it is listed

A little bit of oversight from me it appears. If you `pip install jax`, it does NOT install `jaxlib`, but then complains about it right away if you try to import it. Which is weird because `jaxlib` is available on PyPI. Do you think it would be better to pin to `jax >= 0.4.4`?

**Comment by mattwthompson** [comment]
> If you pip install jax, it does NOT install jaxlib,

That's ... not at all what I anticipated or expected.

> Do you think it would be better to pin to jax >= 0.4.4?

The packaging/deployment issues I've run into (here and with a separate project last summer) have all involved a version like 0.1.75 being brought in, hence why I added the 0.3.0 constraint. (That and these warts seem to have been fixed around then.) Otherwise, I can't make a good case for either version over the other - if you have any preference that should win out.

0.4.4 is fairly recent (last month https://github.com/conda-forge/jax-feedstock/pull/111), I'm not sure what tail of older versions the library itself is meant to support. JAX [asserts](https://jax.readthedocs.io/en/latest/deprecation.html) that they follow NEP 29 so support for old-ish Python versions shouldn't be an issue. The only pushback I'd offer is that I personally lean against pinning to too-new versions of things so that the solver doesn't fall back to old builds - an opinion based more on superstition than evidence.

**Comment by mikemhenry** [comment]
Grumble, I am not sure which is the better fix. Run constrained seems better since it explicitly says what we want, don't pull in an ancient version of `jax` and doesn't require us to trust that `jax` pulls down `jaxlib` in future versions. Thoughts?

**Comment by mattwthompson** [comment]
I personally prefer the aesthetics of #31 but I expect #32 or something similar to work better

**Comment by mikemhenry** [comment]
@mattwthompson can you add artifact uploads so we can test to see if the package works? https://conda-forge.org/docs/maintainer/conda_forge_yml.html#azure
The only issue I see with #31 is https://github.com/conda-forge/pymbar-feedstock/issues/30#issuecomment-1466928858 seems to indicate it won't work, but if we can download the package we should be able to test

**Comment by Lnaden** [comment]
On Conda, the install correctly gets `jaxlib`, its just the `pip` installs which it does not. Would it be better to only define `jax` as a "requires" and instead do a `run_constrained` on the `jaxlib >= 0.3.0`? That way we get the clean requirements, and constrain to avoid the bug.

**Comment by Lnaden** [comment]
Fixed by #33

**Comment by mattwthompson** [comment]
Yup, not gonna test it but it should be fine now
